core/src/main/java/com/datastrato/gravitino/authorization/AuthorizationManager.java

```diff
@@ -0,0 +1,208 @@
+
+  private final Config config;
+
+  @VisibleForTesting final Cache<CatalogEntity, AuthorizationWrapper> authorizationCache;
+
+  public AuthorizationManager(Config config) {
+    this.config = config;
+
+    long cacheEvictionIntervalInMs = config.get(Configs.AUTHORIZATION_CACHE_EVICTION_INTERVAL_MS);
+    this.authorizationCache =
+        Caffeine.newBuilder()
+            .expireAfterAccess(cacheEvictionIntervalInMs, TimeUnit.MILLISECONDS)
+            .removalListener(
+                (k, v, c) -> {
+                  LOG.info("Closing authorization {}.", k);
+                  ((AuthorizationWrapper) v).close();
+                })
+            .scheduler(
+                Scheduler.forScheduledExecutorService(
+                    new ScheduledThreadPoolExecutor(
+                        1,
+                        new ThreadFactoryBuilder()
+                            .setDaemon(true)
+                            .setNameFormat("authorization-cleaner-%d")
+                            .build())))
+            .build();
+  }
+
+  @Override
+  public void close() {
+    authorizationCache.invalidateAll();
+  }
+
+  public AuthorizationWrapper loadAuthorizationAndWrap(CatalogEntity entity)
+      throws NoSuchAuthorizationException {
+    return authorizationCache.get(entity, this::createAuthorizationWrapper);
+  }
+
+  /** Wrapper class for an Authorization instance and its class loader. */
+  public static class AuthorizationWrapper {
+    private BaseAuthorization authorization;
+
+    private IsolatedClassLoader classLoader;
+
+    public AuthorizationWrapper(BaseAuthorization auth, IsolatedClassLoader classLoader) {
+      this.authorization = auth;
+      this.classLoader = classLoader;
+    }
+
+    public <R> boolean runAuthorizationChain(Function<AuthorizationOperations, R>... functions) {
+      try {
+        return classLoader.withClassLoader(
+            cl -> {
+              for (Function<AuthorizationOperations, R> function : functions) {
+                function.apply(authorization.ops());
+              }
+              return true;
+            });
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @VisibleForTesting
+    public AuthorizationOperations getOps() {
+      return authorization.ops();
+    }
+
+    public void close() {
+      try {
+        classLoader.withClassLoader(
+            cl -> {
+              if (authorization != null) {
+                authorization.close();
+              }
+              authorization = null;
+              return null;
+            });
+      } catch (Exception e) {
+        LOG.warn("Failed to close authorization", e);
+      }
+
+      classLoader.close();
+    }
+  }
+
+  @SafeVarargs
+  public final <R> boolean runAuthorizationChain(
+      CatalogEntity entity, Function<AuthorizationOperations, R>... functions) {
+    return loadAuthorizationAndWrap(entity).runAuthorizationChain(functions);
+  }
```

Comment:
good, we can simplify our code by this design，which is more readable

i think we need to add some comments to inform users that they should ensure idempotency by themselves.

below is our current usage，the logic is complex
```java
for (WorkspaceRole role : WorkspaceRole.values()) {
            String gravitinoRoleName = getGravitinoRoleName(qualifiedName, role);
            final SecurableObject.Type securableObjectType = getSecurableObjectType(miSqlResource.getMiSqlType());
            final List<Privilege> privileges = getSecurableObjectPrivileges(securableObjectType, role);
            try {
                adminClient.createRole(metalake, gravitinoRoleName, null, SecurableObjects.parse(qualifiedName, securableObjectType), privileges);
            } catch (Exception e) {
                log.warn("role already exists, gravitinoRoleName: {}", gravitinoRoleName, e);
            }

            String gravitinoGroupName = getGravitinoGroupName(workspaceId, role.getRoleName());
            try {
                Group group = adminClient.getGroup(metalake, gravitinoGroupName);
            } catch (NoSuchGroupException e) {
                log.warn("group {} not exists, create the group", gravitinoGroupName, e);
                adminClient.addGroup(metalake, gravitinoGroupName);
            }
            adminClient.grantRolesToGroup(metalake, Collections.singletonList(gravitinoRoleName), gravitinoGroupName);
        }
```

Reply:
hi @lw-yang Do you need an authorization chain call? what scene needs this future?

Currently, `runAuthorizationChain()` function only support authorization chain call in the one access control system.

We are consider implement cross multiple access control system chain call. 
for example. 
```
authrorizationManager.chain( rangerHiveAuth.createRole(xxx), rangerTrinoAuth.createRole(xxx), ...)
```

What is your opinion?

Reply:
I think what @lw-yang commented here is unrelated the chain design you mentioned.

Reply:
After reconsideration, I realized I misunderstand this

When creating a Fileset on our date platform, current usage involves two steps
```java
createFileset();
grant();
```

The `grant()` logic is complex, and requires multiple calls to Gravitino's API
1. `getRole` check if the role exist 
2. `createRole` if role not exist
3. `getGroup` check if the group exist
4. `createGroup` if group not exist, and need to call this API for three times (workspace.admin, workspace.dev, workspace.visitor)
5. `grantRolesToGroup` also call this API for three times

Reply:
After reconsideration, I realized I misunderstand this

When creating a Fileset on our date platform, current usage involves two steps
```java
createFileset();
grant();
```

The `grant()` logic is complex, and requires multiple calls to Gravitino's API
1. `getRole` check if the role exist 
2. `createRole` if role not exist
3. `getGroup` check if the group exist
4. `createGroup` if group not exist, and need to call this API for three times (workspace.admin, workspace.dev, workspace.visitor)
5. `grantRolesToGroup` also call this API for three times

so there are 5+ api calls in a grant operation

We need an extension point within Gravitino to extend our custom authorization logic. This logic could be executed automatically after a resource is created. The authorization logic can be implemented as a chain of calls.